File: scala_studying/src/main/scala/cn/qf/streaming/day02/studymyself/ProducerDemo.scala
package cn.qf.streaming.day02.studymyself

import java.util.Properties

import kafka.producer.{KeyedMessage, Partitioner, Producer, ProducerConfig}
import kafka.utils.VerifiableProperties

/**
 * Description：<br/>
 * Copyright (c) ,2019 , Xuefengtao <br/>
 * This program is protected by copyright laws. <br/>
 * Date： 2019年11月13日  
 *
 * @author 陶雪峰
 * @version : 1.0
 */
object ProducerDemo {
  def main(args: Array[String]): Unit = {
    // 定义topic，把数据传到该topic
    val topic = "KafkaSimple"

    // 创建一个配置文件信息类
    val props: Properties = new Properties()
    // 数据在序列化时的编码类型
    props.put("serializer.class", "kafka.serializer.StringEncoder")
    // kafka集群列表
    props.put("metadata.broker.list", "192.168.8.101:9092,192.168.8.102:9092,192.168.8.103:9092")
    // 设置发送数据后是否需要服务端的反馈：0，1，-1
    props.put("request.required.acks", "1")
    // 调用分区器
    props.put("partitioner.class", "cn.qf.streaming.day02.studymyself.CustomPartitioner")
    //    props.put("partitioner.class", "kafka.producer.DefaultPartitioner")

    val config = new ProducerConfig(props)

    // 创建一个生产者实例
    val producer: Producer[String, String] = new Producer(config)

    // 模拟生产一些数据
    for (i <- 1 to 10000){
      val msg = s"$i: Producer send data"
      producer.send(new KeyedMessage[String, String](topic, msg))
    }


  }
}
class CustomPartitioner(props: VerifiableProperties) extends Partitioner{
  override def partition(key: Any, numPartitions: Int): Int = {
    key.hashCode() % numPartitions
  }
}

File: scala_studying/src/main/java/cn/qf/kafka/day18/kafkademo/ProducerDemo.java
package cn.qf.kafka.day18.kafkademo;


import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;

import java.util.Properties;

/**
 * Description：模拟生产者,将数据不断发送到kafka<br/>
 * Copyright (c) ,2019 , Xuefengtao <br/>
 * This program is protected by copyright laws. <br/>
 * Date： 2019年11月13日
 *
 * @author 陶雪峰
 * @version : 1.0
 */


public class ProducerDemo {
    public static void main(String[] args) {
        Properties prop = new Properties();
        //指定kafka的集群列表
        prop.put("bootstrap.servers","192.168.8.101:9092,192.168.8.102:9092,192.168.8.103:9092");
        //对producer的响应方式
        prop.put("request.required.acks", 1);
        //key的序列化类
        prop.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        //对value的序列化类
        prop.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");

        String topic = "test";

        //实例化对象
        //key:存储偏移量offset,可以为空
        //value:存储数据
        KafkaProducer<String, String> producer = new KafkaProducer<String, String>(prop);

        int i = 0;
        int max = 10000;
        while (i <= max) {
            String msg = "hello, " + i;
            producer.send(new ProducerRecord<String, String>(topic, msg));
            i++;
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}
